src/crypto/slow-hash.c

```diff
@@ -59,35 +61,136 @@ extern int aesb_pseudo_round(const uint8_t *in, uint8_t *out, const uint8_t *exp
-  if (variant > 0) \
+  if (variant == 1) \
   { \
     VARIANT1_CHECK(); \
   } \
-  const uint64_t tweak1_2 = variant > 0 ? (state.hs.w[24] ^ (*((const uint64_t*)NONCE_POINTER))) : 0
+  const uint64_t tweak1_2 = (variant == 1) ? (state.hs.w[24] ^ (*((const uint64_t*)NONCE_POINTER))) : 0
+
+#define VARIANT2_INIT() \
+  uint64_t division_result; \
+  uint32_t sqrt_result; \
+  do if (variant >= 2) \
+  { \
+    division_result = 0; \
+    sqrt_result = 0; \
+  } while (0)
+
+#define VARIANT2_SHUFFLE_SSE2(base_ptr, offset) \
+  do if (variant >= 2) \
+  { \
+    const __m128i chunk1 = _mm_load_si128((__m128i *)((base_ptr) + ((offset) ^ 0x10))); \
+    const __m128i chunk2 = _mm_load_si128((__m128i *)((base_ptr) + ((offset) ^ 0x20))); \
+    const __m128i chunk3 = _mm_load_si128((__m128i *)((base_ptr) + ((offset) ^ 0x30))); \
+    _mm_store_si128((__m128i *)((base_ptr) + ((offset) ^ 0x10)), _mm_shufflelo_epi16(_mm_shuffle_epi32(chunk3, _MM_SHUFFLE(2, 0, 3, 1)), _MM_SHUFFLE(3, 1, 2, 0))); \
+    _mm_store_si128((__m128i *)((base_ptr) + ((offset) ^ 0x20)), _mm_shufflelo_epi16(_mm_shuffle_epi32(chunk1, _MM_SHUFFLE(0, 3, 1, 2)), _MM_SHUFFLE(0, 3, 1, 2))); \
+    _mm_store_si128((__m128i *)((base_ptr) + ((offset) ^ 0x30)), _mm_shufflelo_epi16(_mm_shuffle_epi32(chunk2, _MM_SHUFFLE(3, 1, 2, 0)), _MM_SHUFFLE(2, 0, 3, 1))); \
+  } while (0)
+
+#define VARIANT2_SHUFFLE_PORTABLE(base_ptr, offset) \
+  do if (variant >= 2) \
+  { \
+    uint32_t* chunk1 = (uint32_t*)((base_ptr) + ((offset) ^ 0x10)); \
+    uint32_t* chunk2 = (uint32_t*)((base_ptr) + ((offset) ^ 0x20)); \
+    uint32_t* chunk3 = (uint32_t*)((base_ptr) + ((offset) ^ 0x30)); \
+    \
+    uint32_t chunk1_old[4]; \
+    memcpy(chunk1_old, chunk1, sizeof(chunk1_old)); \
+    \
+    chunk1[0] = (chunk3[1] & 0x0000FFFFUL) | (chunk3[3] << 16); \
+    chunk1[1] = (chunk3[3] & 0xFFFF0000UL) | (chunk3[1] >> 16); \
+    chunk1[2] = chunk3[0]; \
+    chunk1[3] = chunk3[2]; \
+    \
+    chunk3[0] = (chunk2[0] >> 16) | (chunk2[2] & 0xFFFF0000UL); \
+    chunk3[1] = (chunk2[2] << 16) | (chunk2[0] & 0x0000FFFFUL); \
+    chunk3[2] = chunk2[1]; \
+    chunk3[3] = chunk2[3]; \
+    \
+    chunk2[0] = (chunk1_old[2] & 0xFFFF0000UL) | (chunk1_old[1] & 0x0000FFFFUL); \
+    chunk2[1] = (chunk1_old[1] >> 16) | (chunk1_old[2] << 16); \
+    chunk2[2] = chunk1_old[3]; \
+    chunk2[3] = chunk1_old[0]; \
+  } while (0)
+
+#define VARIANT2_INTEGER_MATH_DIVISION_STEP(b, ptr) \
+  ((uint64_t*)(b))[1] ^= division_result ^ sqrt_result; \
+  { \
+    const uint64_t dividend = ((uint64_t*)(ptr))[1]; \
+    const uint32_t divisor = ((uint32_t*)(ptr))[0] | 0x80000001UL; \
+    division_result = ((uint32_t)(dividend / divisor)) + \
+                     (((uint64_t)(dividend % divisor)) << 32); \
+  } \
+  const int64_t sqrt_input = (((uint64_t*)(ptr))[0] + division_result) >> 16;
+
+#define VARIANT2_INTEGER_MATH_SSE2(b, ptr) \
+  do if (variant >= 2) \
+  { \
+    VARIANT2_INTEGER_MATH_DIVISION_STEP(b, ptr) \
```

Comment:
Maybe pedantic, but it'd be prefereable to lose the ending ; in VARIANT2_INTEGER_MATH_DIVISION_STEP so that statements like the above look like clean C :)